dynamic_rnn time_major=False transpose fixed for inputs of rank>3

ankush-me opened the pull request:
Fix for [dynamic_rnn](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L682) for inputs of rank > 3 when time_major=False, as discussed in issue #3882.

ankush-me committed:
dynamic_rnn time_major=False transpose fixed for inputs of rank>3

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ankush-me commented:
CLA signed.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -785,7 +785,7 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
   if not time_major:
     # (B,T,D) => (T,B,D)
-    flat_input = tuple(array_ops.transpose(input_, [1, 0, 2])
+    flat_input = tuple(array_ops.transpose(input_, [1, 0]+range(2,input_.get_shape().ndims))
```
sadly this won't work if input_ has unknown rank.  you'll have to do something like:

input_rank = input_.get_shape().ndims
if input_rank is None:
  transpose_indices = array_ops.concat(0, ([1, 0], array_ops.range(2, array_ops.rank(input_)))
else:
  transpose_indices = [1, 0] + range(2, input_rank)

ankush-me replied:
Thanks. Got it.

ebrevdo commented:
(please add unit tests with both known and unknown input rank)

On Sat, Oct 22, 2016 at 3:58 PM, Ankush Gupta <notifications@github.com>
wrote:

> CLA signed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-255559000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim3J3Z6S_C07lpz7k4u0zm7Kxch8iks5q2pUFgaJpZM4Kd_s4>
> .
>

ankush-me commented:
@ebrevdo -- yes, will do.

ebrevdo commented:
Any update on this?

ankush-me committed:
dynamic rnn bug fix

ankush-me commented:
@ebrevdo : Since dynamic_rnn calls _dynamic_rnn_loop which requires that all inputs have [well defined shapes at the graph construction time](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L917), it is not clear if dynamic_rnn should handle this case where the rank is unknown. 

I patched it as suggested, but then there is no good way of unit-testing it for inputs with unknown shapes, due to the above requirement. What do you suggest?

ebrevdo commented:
You can probably remove that check. The rnncells themselves can check for
static shape info if needed.

On Nov 12, 2016 5:55 PM, "Ankush Gupta" <notifications@github.com> wrote:

> @ebrevdo <https://github.com/ebrevdo> : Since dynamic_rnn calls
> _dynamic_rnn_loop which requires that all inputs have well defined shapes
> at the graph construction time
> <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py#L917>,
> it is not clear if dynamic_rnn should handle this case where the rank is
> unknown.
>
> I patched it as suggested, but then there is no good way of unit-testing
> it for inputs with unknown shapes, due to the above requirement. What do
> you suggest?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-260161255>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim4wgEUruDdhSPsm4NVouJh7dcH-Uks5q9m31gaJpZM4Kd_s4>
> .
>

ankush-me committed:
removed shape checks in dynamic_rnn and dynamic_rnn_loop for unit-testing -- seem to work

ankush-me committed:
added dynamic-rnn unit tests

ankush-me commented:
I removed the shape checks (a2ff968) and added some unit tests (68c874a) -- both with known and unknown ranks of the inputs and outputs at the graph construction time and it seems to work. However, again this required taking out a lot of the shape-checking -- so not sure if we want to keep it as it is, or bring the checks back in. The unit tests would error out if the shape checks are included, hence the tests were not added to to the rnn_test.py file.

ebrevdo commented:
The only shape checks you need to remove are the lines:

 if not shape[2:].is_fully_defined():
      raise ValueError(
          "Input size (depth of inputs) must be accessible via shape
inference,"
          " but saw value None.")

and these lines (redundant and buggy?):

for input_ in input_shape:
      if input_[1].get_shape() != batch_size.get_shape():
        raise ValueError("All inputs should have the same batch size")

*everything else can stay*


On Tue, Nov 15, 2016 at 10:42 AM, Ankush Gupta <notifications@github.com>
wrote:

> I removed the shape checks (a2ff968
> <https://github.com/tensorflow/tensorflow/commit/a2ff968643512b50ea074edcd0561337dad427e2>)
> and added some unit tests (68c874a
> <https://github.com/tensorflow/tensorflow/commit/68c874a507498863b14e06af05fa512d97d38455>)
> -- both with known and unknown ranks of the inputs and outputs at the graph
> construction time and it seems to work. However, again this required taking
> out a lot of the shape-checking -- so not sure if we want to keep it as it
> is, or bring the checks back in. The unit tests would error out if the
> shape checks are included, hence the tests were not added to to the
> rnn_test.py file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-260728417>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtimzqlqR0frGBjKmxwZQJTxNBpHFLCks5q-f0CgaJpZM4Kd_s4>
> .
>

ankush-me commented:
Correct regarding [this](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L830); should not have been commented out.

But everything else, due to [line 923](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L923):
```
const_time_steps, const_batch_size = inputs_got_shape[0].as_list()[:2]
```
where state shape is required, needs to be taken out (?), i.e.: 

1. Lines [930](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L930) - [938](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L938):
	```
    got_time_steps = shape[0].value
    got_batch_size = shape[1].value
    if const_time_steps != got_time_steps:
      raise ValueError(
          "Time steps is not the same for all the elements in the input in a "
          "batch.")
    if const_batch_size != got_batch_size:
      raise ValueError(
          "Batch_size is not the same for all the elements in the input.")
	```

2. And, lines [1026](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L1026) - [1030](https://github.com/tensorflow/tensorflow/blob/a2ff968643512b50ea074edcd0561337dad427e2/tensorflow/python/ops/rnn.py#L1030)
	```
	# Restore some shape information
	for output, output_size in zip(final_outputs, flat_output_size):
		shape = _state_size_with_prefix(output_size, prefix=[const_time_steps, const_batch_size])
		output.set_shape(shape)
    ```

martinwicke commented:
@ebrevdo, @ankush-me whose turn?

ankush-me commented:
@martinwicke: @ebrevdo, I guess.

martinwicke commented:
He's on vacation, many people are. I suppose this will have to wait a little longer.

drpngx commented:
Happy New Year everyone!

@ankush-me could you sync and push again?

ankush-me committed:
local

ankush-me committed:
Merge remote-tracking branch 'tf_google/master' into dynamic-rnn-fix

ankush-me commented:
@drpngx Sorry for the delay. I've synced with master.

drpngx commented:
Jenkins, test this please.

drpngx commented:
This fails tests. Could you fix?

rmlarsen commented:
@tensorflow-jenkins test this please

vrv commented:
@ankush-me will you have time to look at this again any time soon?  I'll close by the end of the week if not.  Thanks again for your contribution.

ankush-me commented:
@vrv: there are some assumptions about shapes being fully defined at graph construction time in ` _dynamic_rnn_loop`, as highlighted in my comment posed on Nov 15, 2016 above. Due to these assumptions, I cannot write unit tests with inputs with unknown shapes at graph construction time, as is desired by @ebrevdo. Can I take out these checks (as pointed out in the comment above)? Please let me know. Thanks.

ebrevdo commented:
Can you modify the checks to only run it the static values are available?

On Feb 13, 2017 1:50 PM, "Ankush Gupta" <notifications@github.com> wrote:

> @vrv <https://github.com/vrv>: there are some assumptions about shapes
> being fully defined at graph construction time in _dynamic_rnn_loop, as
> highlighted in my comment posed on Nov 15, 2016 above. Due to these
> assumptions, I cannot write unit tests with inputs with unknown shapes at
> graph construction time, as is desired by @ebrevdo
> <https://github.com/ebrevdo>. Can I take out these checks (as pointed out
> in the comment above)? Please let me know. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-279533867>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim6KJgbAKLvSO99hiVBmtGtVFcE2Fks5rcNAegaJpZM4Kd_s4>
> .
>

vrv commented:
@ankush-me any updates?  I'm going to close by the end of the day to keep the PR list small, but feel free to send a new PR if you get around to fixing this.

ankush-me commented:
@vrv I should find time over the weekend to fix this. If you can wait till Monday, that'll be great.

ankush-me committed:
Merge remote-tracking branch 'tf_google/master' into dynamic-rnn-fix

ankush-me committed:
removed shape checks + added unit tests in kernel_tests/rnn_test.py

ankush-me commented:
Done: removed the unnecessary shape checks and added unit tests for input tensors of rank > 3 in kernel_tests/rnn_test.py for both known and unknown static rank. Can you please test this?

drpngx commented:
Jenkins, test this please.

ankush-me committed:
restored build_config.bzl

drpngx commented:
Jenkins, test this please.

Are you done pushing?

ankush-me commented:
Yes.

ankush-me committed:
bug fix

ankush-me committed:
all tests pass locally

ankush-me commented:
Fixed bugs. Tests seem to pass locally. Can you please test this again?

ebrevdo commented on tensorflow/python/kernel_tests/rnn_test.py:
```diff
@@ -33,6 +33,8 @@
 from tensorflow.core.protobuf import config_pb2
 from tensorflow.python.client import session
 from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import tensor_shape
```
imports should be alphabetically ordered.

ebrevdo commented on tensorflow/python/kernel_tests/rnn_test.py:
```diff
@@ -63,6 +65,40 @@ def __call__(self, input_, state, scope=None):
     return (input_ + 1, state + 1)
 
 
+class DummyMultiDimensionalLSTM(rnn_cell_impl._RNNCell):
```
append to line: "  # pylint: disable=protected-access"

ebrevdo commented on tensorflow/python/kernel_tests/rnn_test.py:
```diff
@@ -63,6 +65,40 @@ def __call__(self, input_, state, scope=None):
+
+    Args:
+      dims: tuple that contains the dimensions of the output of the cell,
+      without including 'Time' or 'Batch' dimensions.
```
Raises: section.

ebrevdo commented on tensorflow/python/kernel_tests/rnn_test.py:
```diff
@@ -554,6 +590,52 @@ def benchmarkStaticUnrollStateConcatVsStateTuple(self):
                 iters=20,
                 wall_time=t_dt)
 
+class DynamicRNNInputND(test.TestCase):
```
this is a test, the class name should end with "Test"

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -23,6 +23,7 @@
 from tensorflow.python.framework import dtypes
 from tensorflow.python.framework import ops
 from tensorflow.python.framework import tensor_shape
+
```
remove unnecessary whitespace

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -485,8 +486,18 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
+      x_rank = x.get_shape().ndims
+      if x_rank is None:
+        x_rank = array_ops.rank(x)
+        indices = array_ops.concat([[1,0],math_ops.range(2,x_rank)],0)
```
spaces after every comma, here and elsewhere.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -485,8 +486,18 @@ def dynamic_rnn(cell, inputs, sequence_length=None, initial_state=None,
+        indices = [1,0] + range(2,x_rank)
+      return indices
+
+    flat_input = tuple(array_ops.transpose(input_,
```
newline / indentation are weird; please fix while keeping w/in 80 char limit.

ankush-me replied:
Not exactly sure if I understand this correctly. Can you please check again?

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -550,8 +561,8 @@ def _assert_has_shape(x, shape):
-      flat_output = [array_ops.transpose(output, [1, 0, 2])
-                     for output in flat_output]
+      flat_output = [array_ops.transpose(output,
+                     _get_transpose_indices(output)) for output in flat_output]
```
ditto

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -611,22 +622,7 @@ def _dynamic_rnn_loop(cell,
-          " but saw value None.")
-    got_time_steps = shape[0].value
-    got_batch_size = shape[1].value
-    if const_time_steps != got_time_steps:
```
removing these things is weird; we're just checking consistency across multiple input tensors.  why do you need to remove these lines?  perhaps these ifs should only happen if const_time_steps and got_time_steps are both not None?

ankush-me replied:
I've put these back in. 
`const_time_steps` and `const_batch_size` are taken from the first input; not clear if that is a good idea, if the first input has these values as `None`.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -611,22 +622,7 @@ def _dynamic_rnn_loop(cell,
-  const_time_steps, const_batch_size = inputs_got_shape[0].as_list()[:2]
-
-  for shape in inputs_got_shape:
-    if not shape[2:].is_fully_defined():
```
i suppose this should be: "if shape.ndims is None or shape.ndims < 2:"

ankush-me replied:
Since we want this to run even when the static rank is unknown, these checks should be removed, right? As also discussed [here](https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-260748149).

ankush-me commented:
@ebrevdo Thanks for the review! Will make these changes soon.

ankush-me committed:
Merge remote-tracking branch 'tf_google/master' into dynamic-rnn-fix

ankush-me committed:
changes from review

ankush-me committed:
checking shapes if available

ankush-me committed:
bug fix

drpngx commented:
Jenkins, test this please.

ebrevdo commented:
We actually have an internal piece of code we're gonna move over here to
fix this problem; if that's OK:

https://github.com/tensorflow/tensorflow/blob/34df7ee6f5a8f931d2433dc7e6e739bc880a35ea/tensorflow/contrib/seq2seq/python/ops/decoder.py#L41

So perhaps we can hold off on this PR until that code is moved internally?
Because then it can be used here and in the dynamic decoder.

On Wed, Mar 1, 2017 at 3:44 PM, drpngx <notifications@github.com> wrote:

> Jenkins, test this please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-283509050>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim0tyX4lD-RtlngZzfuU2khTRC1Wnks5rhgLYgaJpZM4Kd_s4>
> .
>

ankush-me commented:
Ok.

ankush-me committed:
small fix for Python 3

marcociccone commented:
any update on this PR? When do you think it's going to be merged? 
for reference: @carlthome

ebrevdo commented:
I just pushed a fix internally; it should show up on master
today/tomorrow.  It's safe to close this PR.

On Thu, Apr 6, 2017 at 1:57 AM, Marco Ciccone <notifications@github.com>
wrote:

> any update on this PR? When do you think it's going to be merged?
> for reference: @carlthome <https://github.com/carlthome>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/5142#issuecomment-292111462>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtim1vInJjXe2cpiF8tBF8uARZqyqDjks5rtKjpgaJpZM4Kd_s4>
> .
>

carlthome commented:
Wonderful! Hope this makes it onto PyPI ASAP! Thanks for notifying @marcociccone.
